google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +400,54 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        outstandingReceipts.put(
+            message.getAckId(), new ReceiptCompleteData(outstandingMessage, false));
+      } else {
+        outstandingBatch.add(outstandingMessage);
+      }
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  synchronized void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      ReceiptCompleteData receiptCompleteData = outstandingReceipts.get(ackRequestData.getAckId());
+      OutstandingMessage outstandingMessage = receiptCompleteData.getOutstandingMessage();
+      // Setting to true means that the receipt is complete
+      receiptCompleteData.setReceiptComplete(true);
+      outstandingReceipts.put(ackRequestData.getAckId(), receiptCompleteData);
+      List<OutstandingMessage> completedReceipts = new ArrayList<>();
+      if (pendingMessages.putIfAbsent(
```

Comment:
We don't want to add this message to pendingMessages unless we are actually going to deliver this message (which we won't know until we check if the receipt is complete.

We should make sure we have test coverage for the case where outstanding acks are completed out of order.

Reply:
when we call this function of notifyAckSuccess() it should mean that the receipt is complete though. thats we set the outstanding message's receipt complete to true. Do you think we should move this logic down to line 434 (when we actually send the messages out to process the batch?

Reply:
thinking about it actually, we are adding to the pending messages, and then going inside to check whether to process the batch of receipts or not. this logic may be opposite of what we want. we want to first process the receipt as fully complete (aka we are ready to process it), and then add it to the pendingMessages correct? will switch the logic.